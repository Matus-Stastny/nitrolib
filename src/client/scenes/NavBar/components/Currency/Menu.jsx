// @flow strict
import * as React from "react";
import * as R from "ramda";
import styled from "styled-components";

import * as fetchedContext from "client/services/fetched/context";
import * as intlContext from "client/services/intl/context";
import type { Currency, Currencies } from "client/records/Currency";
import type { GeoDataQueryResponse } from "client/components/GeoData/__generated__/GeoDataQuery.graphql";
import getGeoIPCountry from "client/services/session/getGeoIPCountry";
import { BORDER_RADIUS } from "client/consts/layout";
import mq from "client/services/utils/mediaQuery";
import { brandDefault } from "client/records/Brand";
import getRecommended from "./services/getRecommended";
import CurrencyList from "./CurrencyList";

type Props = {|
  current: Currency,
  available: Currencies,
  setCurrency: string => void,
  geoIP: ?GeoDataQueryResponse,
  hide: () => void,
|};

const Container = styled.div`
  position: absolute;
  right: 0;
  width: calc(100% - 40px);
  max-height: calc(100vh - 200px);
  margin 0 20px;
  box-sizing: border-box;
  padding: 15px 12px;
  overflow-y: scroll;
  box-shadow: 0 0 10px 5px rgba(0, 0, 0, .1);
  border-radius: ${BORDER_RADIUS}px;
  background: ${props => props.theme.colors.white};
  color: ${props => props.theme.colors["grey-800"]};

  ${mq.gtTablet`
    width: 500px;
    padding: 20px;
    margin: 0;
    max-height: calc(100vh - 60px);
    overflow-y: auto;
  `}

  ${mq.tablet`
    right: 0; /* TODO: RTL */
    left: inherit; /* TODO: RTL */
  `}

  ${mq.gtDesktop`
    left: -290px; /* TODO: RTL */
    right: inherit; /* TODO: RTL */
  `}
`;

const Recommended = styled.div`
  position: relative;
  width: 100%;
  padding-bottom: 13px;
  margin-bottom: 12px;

  &:after {
    position: absolute;
    bottom: 0;
    left: 0;
    display: block;
    content: "";
    width: calc(100% - 10px);
    height: 1px;
    margin: 0 5px;
    background: ${props => props.theme.colors["grey-200"]};
  }
`;

Recommended.defaultProps = {
  theme: brandDefault.theme,
};

class Menu extends React.Component<Props> {
  setCurrency = (code: string) => {
    // TODO: logging
    this.props.setCurrency(code);
    this.props.hide();
  };

  render() {
    const { current, available, geoIP } = this.props;

    return (
      <fetchedContext.Consumer>
        {/* istanbul ignore next */
        ({ countries }) => (
          <intlContext.Consumer>
            {({ language }) => {
              const recommended = getRecommended(
                getGeoIPCountry(geoIP, countries),
                language,
                available,
              );

              return (
                <Container>
                  {!R.isEmpty(recommended) && (
                    <Recommended>
                      <CurrencyList
                        currencies={recommended}
                        active={current}
                        setCurrency={this.setCurrency}
                      />
                    </Recommended>
                  )}
                  <CurrencyList
                    currencies={available}
                    active={current}
                    setCurrency={this.setCurrency}
                  />
                </Container>
              );
            }}
          </intlContext.Consumer>
        )}
      </fetchedContext.Consumer>
    );
  }
}

export default Menu;
