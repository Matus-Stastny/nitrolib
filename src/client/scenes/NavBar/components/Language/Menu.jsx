// @flow strict
import * as React from "react";
import * as R from "ramda";
import styled from "styled-components";

import * as fetchedContext from "client/services/fetched/context";
import Flex from "client/primitives/Flex";
import Text from "client/components/Text";
import mq from "client/services/utils/mediaQuery";
import type { ThemeProps } from "client/records/Brand";
import type { Languages } from "client/records/Languages";
import LanguageName from "./LanguageName";
import ContinentName from "./ContinentName";
import { getLanguageWrapperHeight, getLanguageWrapperWidth } from "./services/menu";

const MenuWrapper = styled.div`
  background-color: white;
  position: absolute;
  top: 45px;
  right: 20px;
  left: auto;
  display: flex;
  border-radius: 3px;
  box-shadow: 0 0 10px 5px rgba(0, 0, 0, 0.1);
  ${mq.ltTablet`
    left: 20px;
  `};
`;

const ContinentList = styled.div`
  min-width: 200px;
  font-size: 12px;
  font-weight: 500;
  ${mq.ltTablet`
    display: none;
  `};
`;

type ActiveProps = ThemeProps & {|
  active: boolean,
|};

const ContinentItem = styled(Flex)`
  background-color: ${({ theme, active }: ActiveProps) => active && theme.colors["neutral-100"]};
  color: ${({ theme, active }: ActiveProps) => active && theme.colors["primary-600"]};
  padding: 14px 16px;
  line-height: 16px;
  cursor: pointer;
  &:hover {
    background-color: ${({ theme }: ActiveProps) => theme.colors["neutral-100"]};
    color: ${({ theme }: ActiveProps) => theme.colors["primary-600"]};
  }
`;

const LanguageList = styled.div`
  padding: 10px;
  line-height: 20px;
  box-sizing: border-box;
  overflow-y: auto;
  overflow-x: hidden;
  max-height: calc(100vh - 60px);
  ${mq.gtTablet`
    width: 560px;
  `};
  ${mq.ltTablet`
    padding: 15px 12px;
  `};
`;

type SizeProps = {|
  width: number,
  height: number,
|};

const LanguageListWrapper = styled.div`
  width: ${({ width }: SizeProps) => width};
  height: ${({ height }: SizeProps) => height};
  max-height: 682px;
  ${mq.gtTablet`
    display: flex;
    flex-direction: column;
    flex-wrap: wrap;
  `};
`;

const LanguageItem = styled.div`
  background-color: ${({ theme, active }: ActiveProps) => active && theme.colors["primary-600"]};
  color: ${({ theme, active }: ActiveProps) => active && theme.colors.white};
  width: 33%;
  height: 26px;
  display: inline-flex;
  align-items: center;
  border-radius: 3px;
  padding-left: 5px;
  margin-bottom: 5px;
  box-sizing: border-box;
  cursor: pointer;
  &:hover {
    background-color: ${({ theme }: ActiveProps) => theme.colors["neutral-200"]};
  }
  ${mq.ltTablet`
    width: 100%;
  `};
`;

type State = {|
  continent: string,
|};

export default class Menu extends React.Component<{}, State> {
  state = {
    continent: "",
  };

  handleContinent = (continent: string) => {
    this.setState({ continent });
  };

  changeLanguage = (languageId: string) => {
    const loc = window.location;
    const parts = loc.pathname.split("/");
    const restOfUrl = parts.slice(2).join("/");

    window.location.assign(`${loc.origin}/${languageId}/${restOfUrl}${loc.search}`);
  };

  filterLanguages = (languages: Languages) => {
    const { continent } = this.state;

    return R.values(languages).filter(
      language => continent === "" || continent === language.continent,
    );
  };

  render() {
    const { continent } = this.state;

    return (
      <fetchedContext.Consumer>
        {fetched => {
          const filteredLanguages = this.filterLanguages(fetched.brandLanguage.languages);

          return (
            <MenuWrapper>
              <ContinentList>
                <ContinentItem onClick={() => this.handleContinent("")}>
                  <Text t={__("common.languages_all")} />
                </ContinentItem>
                {fetched.brandLanguage.continents.map(item => (
                  <ContinentItem
                    onClick={() => this.handleContinent(item)}
                    key={item}
                    active={item === continent}
                  >
                    <ContinentName id={item} />
                  </ContinentItem>
                ))}
              </ContinentList>
              <LanguageList>
                <LanguageListWrapper
                  height={getLanguageWrapperHeight(filteredLanguages)}
                  width={getLanguageWrapperWidth(filteredLanguages)}
                >
                  {filteredLanguages.map(language => (
                    <LanguageItem
                      key={language.id}
                      onClick={() => this.changeLanguage(language.id)}
                    >
                      <LanguageName language={language} key={language.id} />
                    </LanguageItem>
                  ))}
                </LanguageListWrapper>
              </LanguageList>
            </MenuWrapper>
          );
        }}
      </fetchedContext.Consumer>
    );
  }
}
