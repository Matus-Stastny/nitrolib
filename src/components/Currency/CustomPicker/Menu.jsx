// @flow strict
import * as React from "react";
import * as R from "ramda";
import styled, { css } from "styled-components";

import type { Currency } from "../../../records/Currency";
import mq from "../../../styles/mediaQuery";
import { themeDefault } from "../../../records/Theme";
import type { ThemeProps } from "../../../records/Theme";
import CurrencyList from "./CurrencyList";

type Props = {|
  current: Currency,
  available: Currency[],
  recommended: Currency[],
  onSetCurrency: (code: string) => void,
|};

const Container = styled.div`
  position: absolute;
  right: 0;
  width: calc(100% - 40px);
  max-height: calc(100vh - 200px);
  margin: 0 20px;
  box-sizing: border-box;
  padding: 15px 12px;
  overflow-y: scroll;
  box-shadow: 0 0 10px 5px rgba(0, 0, 0, .1);
  border-radius: 3px;
  background: ${({ theme }: ThemeProps) => theme.orbit.paletteWhite};
  color: ${({ theme }: ThemeProps) => theme.orbit.paletteInkNormal};

  ${mq.gtTablet(css`
    width: 500px;
    padding: 20px;
    margin: 0;
    max-height: calc(100vh - 60px);
    overflow-y: auto;
  `)}

  ${mq.tablet(css`
    right: 0; /* TODO: RTL */
    left: inherit; /* TODO: RTL */
  `)}

  ${mq.gtDesktop(css`
    left: -290px; /* TODO: RTL */
    right: inherit; /* TODO: RTL */
  `)}
`;

Container.defaultProps = {
  theme: themeDefault,
};

const Recommended = styled.div`
  position: relative;
  width: 100%;
  padding-bottom: 13px;
  margin-bottom: 12px;

  &:after {
    position: absolute;
    bottom: 0;
    left: 0;
    display: block;
    content: "";
    width: calc(100% - 10px);
    height: 1px;
    margin: 0 5px;
    background: ${({ theme }: ThemeProps) => theme.orbit.paletteCloudNormal};
  }
`;

Recommended.defaultProps = {
  theme: themeDefault,
};

const Menu = ({ current, available, recommended, onSetCurrency }: Props) => (
  <Container>
    {!R.isEmpty(recommended) && (
      <Recommended>
        <CurrencyList list={recommended} active={current} onSetCurrency={onSetCurrency} />
      </Recommended>
    )}
    <CurrencyList list={available} active={current} onSetCurrency={onSetCurrency} />
  </Container>
);

export default Menu;
