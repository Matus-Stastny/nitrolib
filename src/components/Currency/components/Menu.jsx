// @flow strict
import * as React from "react";
import * as R from "ramda";
import styled, { css } from "styled-components";

import type { Currency } from "../../../records/Currency";
import mq from "../../../styles/mq";
import * as rtl from "../../../styles/rtl";
import { themeDefault } from "../../../records/Theme";
import type { ThemeProps } from "../../../records/Theme";
import * as MODALS from "../../../consts/modals";
import type { Modal as ModalType } from "../../../consts/modals";
import CurrencyList from "./CurrencyList";

const Container = styled.div`
  position: absolute;
  ${rtl.right}: 0;
  top: 50px;
  width: calc(100% - 40px);
  max-height: calc(100vh - 200px);
  margin: 0 20px;
  box-sizing: border-box;
  padding: 15px 12px;
  overflow-y: scroll;
  box-shadow: 0 0 10px 5px rgba(0, 0, 0, .1);
  border-radius: ${({ theme }: ThemeProps) => theme.orbit.borderRadiusNormal};
  background: ${({ theme }: ThemeProps) => theme.orbit.paletteWhite};

  ${mq.gtTablet(css`
    width: 500px;
    padding: 20px;
    margin: 0;
    max-height: calc(100vh - 60px);
    overflow-y: auto;
  `)}

  ${mq.tablet(css`
    ${rtl.right}: ${({ positionMenuTablet }) => positionMenuTablet}px;
    ${rtl.left}: inherit;
  `)}

  ${mq.gtDesktop(css`
    ${rtl.left}: inherit;
    ${rtl.right}: ${({ positionMenuDesktop }) => positionMenuDesktop}px;
  `)}
`;

Container.defaultProps = {
  theme: themeDefault,
};

const Recommended = styled.div`
  position: relative;
  width: 100%;
  padding-bottom: 13px;
  margin-bottom: 12px;

  &:after {
    position: absolute;
    bottom: 0;
    ${rtl.left}: 0;
    display: block;
    content: "";
    width: calc(100% - 10px);
    height: 1px;
    margin: 0 5px;
    background: ${({ theme }: ThemeProps) => theme.orbit.paletteCloudNormal};
  }
`;

Recommended.defaultProps = {
  theme: themeDefault,
};

type Props = {|
  current: Currency,
  available: Currency[],
  recommended: Currency[],
  onChange: (code: string) => void,
  positionMenuTablet?: number,
  positionMenuDesktop?: number,
  onSetModal?: (modal: ModalType) => void,
|};

class Menu extends React.Component<Props> {
  componentDidMount() {
    const { onSetModal } = this.props;
    if (onSetModal) {
      onSetModal(MODALS.CURRENCY_MENU);
    }
  }

  componentWillUnmount() {
    const { onSetModal } = this.props;
    if (onSetModal) {
      onSetModal(MODALS.NONE);
    }
  }

  render() {
    const {
      current,
      available,
      recommended,
      onChange,
      positionMenuDesktop,
      positionMenuTablet,
    } = this.props;
    return (
      <Container
        positionMenuDesktop={positionMenuDesktop}
        positionMenuTablet={positionMenuTablet}
        data-test="CurrencySwitcher-List"
      >
        {!R.isEmpty(recommended) && (
          <Recommended>
            <CurrencyList list={recommended} active={current} onSetCurrency={onChange} />
          </Recommended>
        )}
        <CurrencyList list={available} active={current} onSetCurrency={onChange} />
      </Container>
    );
  }
}

export default Menu;
