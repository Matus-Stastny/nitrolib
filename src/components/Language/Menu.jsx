// @flow strict
import * as React from "react";
import * as R from "ramda";
import styled, { css } from "styled-components";

import * as fetchedContext from "../../services/fetched/context";
import Text from "../Text";
import type { ThemeProps } from "../../records/Brand";
import type { Languages } from "../../records/Languages";
import mq from "../../styles/mediaQuery";
import Flex from "../../primitives/Flex";
import LanguageName from "./LanguageName";
import ContinentName from "./ContinentName";
import { getLanguageWrapperHeight, getLanguageWrapperWidth } from "./services/menu";

const MenuWrapper = styled.div`
  background-color: white;
  position: absolute;
  top: 45px;
  right: 20px;
  left: auto;
  display: flex;
  border-radius: 3px;
  box-shadow: 0 0 10px 5px rgba(0, 0, 0, 0.1);
  ${mq.ltTablet(css`
    left: 20px;
  `)};
`;

const ContinentList = styled.div`
  min-width: 200px;
  font-size: 12px;
  font-weight: 500;
  ${mq.ltTablet(css`
    display: none;
  `)};
`;

type ActiveProps = ThemeProps & {|
  active: boolean,
|};

const ContinentItem = styled(Flex)`
  background-color: ${({ theme, active }: ActiveProps) => active && theme.colors["neutral-100"]};
  color: ${({ theme, active }: ActiveProps) => active && theme.colors["primary-600"]};
  padding: 14px 16px;
  line-height: 16px;
  cursor: pointer;
  &:hover {
    background-color: ${({ theme }: ActiveProps) => theme.colors["neutral-100"]};
    color: ${({ theme }: ActiveProps) => theme.colors["primary-600"]};
  }
`;

const LanguageList = styled.div`
  padding: 10px;
  line-height: 20px;
  box-sizing: border-box;
  overflow-y: auto;
  overflow-x: hidden;
  max-height: calc(100vh - 60px);
  ${mq.gtTablet(css`
    width: 560px;
  `)};
  ${mq.ltTablet(css`
    padding: 15px 12px;
  `)};
`;

type SizeProps = {|
  width: number,
  height: number,
|};

const LanguageListWrapper = styled.div`
  width: ${({ width }: SizeProps) => width};
  height: ${({ height }: SizeProps) => height};
  max-height: 682px;
  ${mq.gtTablet(css`
    display: flex;
    flex-direction: column;
    flex-wrap: wrap;
  `)};
`;

const LanguageItem = styled.div`
  background-color: ${({ theme, active }: ActiveProps) => active && theme.colors["primary-600"]};
  color: ${({ theme, active }: ActiveProps) => active && theme.colors.white};
  width: 33%;
  height: 26px;
  display: inline-flex;
  align-items: center;
  border-radius: 3px;
  padding-left: 5px;
  margin-bottom: 5px;
  box-sizing: border-box;
  cursor: pointer;
  &:hover {
    background-color: ${({ theme }: ActiveProps) => theme.colors["neutral-200"]};
  }
  ${mq.ltTablet(css`
    width: 100%;
  `)};
`;

type Props = {|
  onChange: (lang: string) => void,
|};

type State = {|
  continent: string,
|};

export default class Menu extends React.Component<Props, State> {
  state = {
    continent: "",
  };

  handleContinent = (continent: string) => {
    this.setState({ continent });
  };

  changeLanguage = (languageId: string) => {
    const { onChange } = this.props;
    const loc = window.location;
    const parts = loc.pathname.split("/");
    const restOfUrl = parts.slice(2).join("/");

    onChange(languageId);
    window.location.assign(`${loc.origin}/${languageId}/${restOfUrl}${loc.search}`);
  };

  filterLanguages = (languages: Languages) => {
    const { continent } = this.state;

    return R.values(languages).filter(
      language => continent === "" || continent === language.continent,
    );
  };

  render() {
    const { continent } = this.state;

    return (
      <fetchedContext.Consumer>
        {fetched => {
          const filteredLanguages = this.filterLanguages(fetched.brandLanguage.languages);

          return (
            <MenuWrapper>
              <ContinentList>
                <ContinentItem onClick={() => this.handleContinent("")}>
                  <Text t={__("common.languages_all")} />
                </ContinentItem>
                {fetched.brandLanguage.continents.map(item => (
                  <ContinentItem
                    onClick={() => this.handleContinent(item)}
                    key={item}
                    active={item === continent}
                  >
                    <ContinentName id={item} />
                  </ContinentItem>
                ))}
              </ContinentList>
              <LanguageList>
                <LanguageListWrapper
                  height={getLanguageWrapperHeight(filteredLanguages)}
                  width={getLanguageWrapperWidth(filteredLanguages)}
                >
                  {filteredLanguages.map(language => (
                    <LanguageItem
                      key={language.id}
                      onClick={() => this.changeLanguage(language.id)}
                    >
                      <LanguageName language={language} key={language.id} />
                    </LanguageItem>
                  ))}
                </LanguageListWrapper>
              </LanguageList>
            </MenuWrapper>
          );
        }}
      </fetchedContext.Consumer>
    );
  }
}
