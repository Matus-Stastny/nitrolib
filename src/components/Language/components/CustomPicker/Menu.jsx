// @flow strict
import * as React from "react";
import styled, { css } from "styled-components";

import { themeDefault } from "../../../../records/Theme";
import type { ThemeProps } from "../../../../records/Theme";
import { getByContinent } from "../../../../records/Languages";
import type { Language } from "../../../../records/Languages";
import mq from "../../../../styles/mediaQuery";
import Flex from "../../../../primitives/Flex";
import Text from "../../../Text";
import { getLanguageWrapperHeight, getLanguageWrapperWidth } from "../../services/menu";
import LanguageName from "./LanguageName";
import ContinentName from "./ContinentName";

const MenuWrapper = styled.div`
  background-color: white;
  position: absolute;
  top: 45px;
  right: 20px;
  left: auto;
  display: flex;
  border-radius: 3px;
  box-shadow: 0 0 10px 5px rgba(0, 0, 0, 0.1);
  ${mq.ltTablet(css`
    left: 20px;
  `)};
  ${mq.gtDesktop(css`
    right: 150px;
  `)};
`;

const ContinentList = styled.div`
  min-width: 200px;
  font-size: 12px;
  font-weight: 500;
  ${mq.ltTablet(css`
    display: none;
  `)};
`;

type ActiveProps = ThemeProps & {|
  active: boolean,
|};

const ContinentItem = styled(Flex)`
  background-color: ${({ theme, active }: ActiveProps) => active && theme.orbit.paletteCloudNormal};
  color: ${({ theme, active }: ActiveProps) => active && theme.orbit.paletteProductNormal};
  padding: 14px 16px;
  line-height: 16px;
  cursor: pointer;
  &:hover {
    background-color: ${({ theme }: ActiveProps) => theme.orbit.paletteCloudNormal};
    color: ${({ theme }: ActiveProps) => theme.orbit.paletteProductNormal};
  }
`;

ContinentItem.defaultProps = {
  theme: themeDefault,
};

type FlatProps = {|
  flat: boolean,
|};

const LanguageList = styled.div`
  padding: 10px;
  line-height: 20px;
  box-sizing: border-box;
  overflow-y: auto;
  overflow-x: hidden;
  max-height: calc(100vh - 60px);
  ${({ flat }: FlatProps) =>
    !flat &&
    mq.gtTablet(css`
      width: 560px;
    `)};
  ${mq.ltTablet(css`
    padding: 15px 12px;
  `)};
`;

type SizeProps = {|
  width: number,
  height: number,
|};

const LanguageListWrapper = styled.div`
  width: ${({ width }: SizeProps) => width}px;
  height: ${({ height }: SizeProps) => height}px;
  max-height: 682px;
  ${mq.gtTablet(css`
    display: flex;
    flex-direction: column;
    flex-wrap: wrap;
  `)};
`;

const LanguageItem = styled.div`
  background-color: ${({ theme, active }: ActiveProps) =>
    active && theme.orbit.paletteProductNormal};
  color: ${({ theme, active }: ActiveProps) => active && theme.orbit.paletteWhite};
  width: 180px;
  height: 26px;
  display: inline-flex;
  align-items: center;
  border-radius: 3px;
  padding-left: 5px;
  margin-bottom: 5px;
  box-sizing: border-box;
  cursor: pointer;
  &:hover {
    background-color: ${({ theme }: ActiveProps) => theme.orbit.paletteCloudNormalHover};
  }
  ${mq.ltTablet(css`
    width: 100%;
  `)};
`;

LanguageItem.defaultProps = {
  theme: themeDefault,
};

type Props = {|
  languages: Language[],
  continents: string[],
  onChange: (lang: string) => void,
  flat: boolean,
|};

type State = {|
  continent: string,
|};

export default class Menu extends React.Component<Props, State> {
  state = {
    continent: "",
  };

  handleContinent = (continent: string) => {
    this.setState({ continent });
  };

  handleChange = (lang: string) => {
    const { onChange } = this.props;

    onChange(lang);
  };

  render() {
    const { continent } = this.state;
    const { flat, languages, continents } = this.props;

    const filteredLanguages = continent === "" ? languages : getByContinent(languages, continent);

    return (
      <MenuWrapper>
        {!flat && (
          <ContinentList>
            <ContinentItem onClick={() => this.handleContinent("")}>
              <Text t={__("common.languages_all")} />
            </ContinentItem>
            {continents.map(item => (
              <ContinentItem
                onClick={() => this.handleContinent(item)}
                key={item}
                active={item === continent}
              >
                <ContinentName id={item} />
              </ContinentItem>
            ))}
          </ContinentList>
        )}
        <LanguageList flat={flat}>
          <LanguageListWrapper
            height={getLanguageWrapperHeight(filteredLanguages, flat)}
            width={getLanguageWrapperWidth(filteredLanguages, flat)}
          >
            {filteredLanguages.map(language => (
              <LanguageItem key={language.id} onClick={() => this.handleChange(language.id)}>
                <LanguageName language={language} key={language.id} />
              </LanguageItem>
            ))}
          </LanguageListWrapper>
        </LanguageList>
      </MenuWrapper>
    );
  }
}
