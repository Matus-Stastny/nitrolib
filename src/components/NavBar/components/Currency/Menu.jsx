// @flow strict
import * as React from "react";
import * as R from "ramda";
import styled, { css } from "styled-components";
import { graphql, createFragmentContainer } from "react-relay";

import type { ThemeProps } from "records/Theme";
import * as fetchedContext from "services/fetched/context";
import * as intlContext from "services/intl/context";
import type { Currency, Currencies } from "records/Currency";
import { themeDefault } from "records/Theme";
import mq from "styles/mediaQuery";
import getRecommended from "./services/getRecommended";
import CurrencyList from "./CurrencyList";
import type { MenuList_list } from "./__generated__/MenuList_list.graphql";
import type { MenuGeo_geo } from "./__generated__/MenuGeo_geo.graphql";

type Props = {|
  current: Currency,
  available: Currencies,
  list: MenuList_list,
  geo: MenuGeo_geo,
  onSetCurrency: (code: string) => void,
  onHide: () => void,
|};

const Container = styled.div`
  position: absolute;
  right: 0;
  width: calc(100% - 40px);
  max-height: calc(100vh - 200px);
  margin: 0 20px;
  box-sizing: border-box;
  padding: 15px 12px;
  overflow-y: scroll;
  box-shadow: 0 0 10px 5px rgba(0, 0, 0, .1);
  border-radius: 3px;
  background: ${({ theme }: ThemeProps) => theme.orbit.paletteWhite};
  color: ${({ theme }: ThemeProps) => theme.orbit.paletteInkNormal};

  ${mq.gtTablet(css`
    width: 500px;
    padding: 20px;
    margin: 0;
    max-height: calc(100vh - 60px);
    overflow-y: auto;
  `)}

  ${mq.tablet(css`
    right: 0; /* TODO: RTL */
    left: inherit; /* TODO: RTL */
  `)}

  ${mq.gtDesktop(css`
    left: -290px; /* TODO: RTL */
    right: inherit; /* TODO: RTL */
  `)}
`;

Container.defaultProps = {
  theme: themeDefault,
};

const Recommended = styled.div`
  position: relative;
  width: 100%;
  padding-bottom: 13px;
  margin-bottom: 12px;

  &:after {
    position: absolute;
    bottom: 0;
    left: 0;
    display: block;
    content: "";
    width: calc(100% - 10px);
    height: 1px;
    margin: 0 5px;
    background: ${({ theme }: ThemeProps) => theme.orbit.paletteCloudNormal};
  }
`;

Recommended.defaultProps = {
  theme: themeDefault,
};

class Menu extends React.Component<Props> {
  handleSetCurrency = (code: string) => {
    // TODO: logging
    const { onSetCurrency, onHide } = this.props;

    onSetCurrency(code);
    onHide();
  };

  render() {
    const { current, available, geo, list } = this.props;

    return (
      <fetchedContext.Consumer>
        {({ countries }) => (
          <intlContext.Consumer>
            {({ language }) => {
              // TODO turn this into an array of common recommended, and a country one
              // then supply to 'CurrencyList' as an optional prop and filter there
              const recommended = getRecommended(
                countries[geo.isoCountryCode || ""],
                language,
                available,
              );

              return (
                <Container>
                  {!R.isEmpty(recommended) && (
                    <Recommended>
                      {/* $FlowIssue */}
                      <CurrencyList
                        list={list}
                        available={recommended}
                        active={current}
                        onSetCurrency={this.handleSetCurrency}
                      />
                    </Recommended>
                  )}
                  {/* $FlowIssue */}
                  <CurrencyList
                    list={list}
                    available={available}
                    active={current}
                    onSetCurrency={this.handleSetCurrency}
                  />
                </Container>
              );
            }}
          </intlContext.Consumer>
        )}
      </fetchedContext.Consumer>
    );
  }
}

export const MenuUnwrapped = Menu;

export default createFragmentContainer(
  Menu,
  graphql`
    fragment MenuList_list on CurrencyDetailConnection {
      ...CurrencyList_list
    }

    fragment MenuGeo_geo on GeoIP {
      isoCountryCode
    }
  `,
);
