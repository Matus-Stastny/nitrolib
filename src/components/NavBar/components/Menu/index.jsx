// @flow strict
import * as React from "react";
import AccountCircle from "@kiwicom/orbit-components/lib/icons/AccountCircle";
import Modal from "@kiwicom/orbit-components/lib/Modal";
import ModalSection from "@kiwicom/orbit-components/lib/Modal/ModalSection";

import Desktop from "../../../Desktop";
import Mobile from "../../../Mobile";
import Text from "../../../Text";
import * as authContext from "../../../../services/auth/context";
import { Consumer as BrandConsumer } from "../../../../services/brand/context";
import Button from "../../primitives/Button";
import Trips from "../Trips";
import Login from "./components/Login";
import ForgotPassword from "./components/ForgotPassword";
import SideNav from "./components/SideNav";
import MenuSpacings from "../../primitives/MenuSpacings";

type Props = {|
  chat: React.Node,
  subscription: React.Node,
  debug?: React.Node,
  onSaveLanguage: (lang: string) => void,
|};

type AuthModal = "myBooking" | "register" | "signIn" | "forgotPassword";

type State = {|
  modalOpen: "" | AuthModal,
|};

export default class Menu extends React.PureComponent<Props, State> {
  state = {
    modalOpen: "",
  };

  handleClose = () => {
    this.setState({ modalOpen: "" });
  };

  handleOpenMyBooking = () => {
    this.setState({ modalOpen: "myBooking" });
  };

  handleOpenRegister = () => {
    this.setState({ modalOpen: "register" });
  };

  handleOpenSignIn = () => {
    this.setState({ modalOpen: "signIn" });
  };

  handleOpenForgotPassword = () => {
    this.setState({ modalOpen: "forgotPassword" });
  };

  render() {
    const { chat, subscription, debug, onSaveLanguage } = this.props;
    const { modalOpen } = this.state;

    return (
      <>
        <authContext.Consumer>
          {({ auth }) =>
            auth === null ? (
              <MenuSpacings>
                <Desktop display="flex">
                  <Button onClick={this.handleOpenMyBooking}>
                    <Text t={__("account.my_bookings_action")} />
                  </Button>
                </Desktop>
                <Mobile display="flex">
                  <Button onClick={this.handleOpenMyBooking} padding="13px 9px">
                    <AccountCircle />
                  </Button>
                </Mobile>
              </MenuSpacings>
            ) : (
              <Trips auth={auth} />
            )
          }
        </authContext.Consumer>
        <SideNav
          chat={chat}
          subscription={subscription}
          debug={debug}
          onOpenRegister={this.handleOpenRegister}
          onOpenSignIn={this.handleOpenSignIn}
          onSaveLanguage={onSaveLanguage}
        />

        {modalOpen !== "" && (
          <Modal onClose={this.handleClose} size="small">
            <ModalSection>
              {modalOpen === "forgotPassword" ? (
                <BrandConsumer>
                  {brand => <ForgotPassword brandId={brand.id} onClose={this.handleClose} />}
                </BrandConsumer>
              ) : (
                <Login
                  open={modalOpen}
                  onCloseSuccess={this.handleClose}
                  onOpenMyBooking={this.handleOpenMyBooking}
                  onOpenRegister={this.handleOpenRegister}
                  onOpenSignIn={this.handleOpenSignIn}
                  onOpenForgotPassword={this.handleOpenForgotPassword}
                />
              )}
            </ModalSection>
          </Modal>
        )}
      </>
    );
  }
}
