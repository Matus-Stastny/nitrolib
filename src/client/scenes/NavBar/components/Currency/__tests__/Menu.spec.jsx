// @flow strict
import * as React from "react";
import { shallow } from "enzyme";

import Menu from "../Menu";

const available = {
  dkk: {
    id: "dkk",
    name: "Danish Krone",
    format: "__price__ kr",
    uncertainFormat: true,
    round: "0",
    enabledOnAffilId: "",
    fallback: "",
    rate: 0.13434,
  },
  gbp: {
    id: "gbp",
    name: "British Pound Sterling",
    format: "£__price__",
    uncertainFormat: false,
    round: "2",
    enabledOnAffilId: "",
    fallback: "",
    rate: 1.14355,
  },
};

const current = {
  id: "gbp",
  name: "British Pound Sterling",
  format: "£__price__",
  uncertainFormat: false,
  round: "2",
  enabledOnAffilId: "",
  fallback: "",
  rate: 1.14355,
};

const geoIP = {
  geoIP: {
    coordinates: {
      lng: 20,
      lat: 30,
    },
    isoCountryCode: "GB",
  },
};

describe("#Currency/Menu", () => {
  test("render", () => {
    const wrapper = shallow(
      <Menu
        available={available}
        current={current}
        geoIP={geoIP}
        setCurrency={jest.fn()}
        hide={jest.fn()}
      />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  test("setCurrency", () => {
    const setCurrency = jest.fn();
    const hide = jest.fn();

    const wrapper = shallow(
      <Menu
        available={available}
        current={current}
        geoIP={geoIP}
        setCurrency={setCurrency}
        hide={hide}
      />,
    );

    wrapper.instance().setCurrency("czk");

    expect(setCurrency).toBeCalledWith("czk");
    expect(hide).toBeCalled();
  });
});
