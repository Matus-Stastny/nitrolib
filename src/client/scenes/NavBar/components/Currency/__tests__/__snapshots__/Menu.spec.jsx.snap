// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#Currency/Menu render 1`] = `<[object Object] />`;
